lib/src/Uncertainty/Algorithm/Sensitivity/HSICEstimatorImplementation.cxx

```diff
@@ -0,0 +1,413 @@
+      SquareMatrix W(computeWeightMatrix(Yp));
+      Scalar HSIC_loc = computeHSICIndex(xdim, Yp, covarianceList_[dim], covarianceList_[inputDimension_], W);
+      if( HSIC_loc > HSIC_obs) count += 1;
+    }
+
+    /* p-value by permutation */
+    PValuesPermutation_[dim] = count * 1.0 / (permutationSize_ + 1) ;
+  }
+}
+
+/** Get the HSIC indices.
+ *  This is not const as it triggers a computation of the indices
+ *  if they are not computed yet.
+ * */
+Point HSICEstimatorImplementation::getHSICIndices()
+{
+  if( HSIC_XY_.getDimension() == 0)
+  {
+    computeIndices();
+  }
+
+  return HSIC_XY_;
+}
+
+/** Get the R2-HSIC indices.
+ *  This is not const as it triggers a computation of the indices
+ *  if they are not computed yet.
+ * */
+Point HSICEstimatorImplementation::getR2HSICIndices()
+{
+  if( R2HSICIndices_.getDimension() == 0)
+  {
+    computeIndices();
+  }
+  return R2HSICIndices_;
+}
+
+/** Get the p-values by permutation.
+ *  This is not const as it triggers a computation of the values
+ *  if they are not computed yet.
+ * */
+Point HSICEstimatorImplementation::getPValuesPermutation()
+{
+  if( PValuesPermutation_.getDimension() == 0)
+  {
+    computePValuesPermutation();
+  }
+  return PValuesPermutation_;
+}
+
+/** Draw the HSIC indices */
+Graph HSICEstimatorImplementation::drawValues(const Point &values, String &title) const
+{
+
+  if (values.getDimension() == 0) throw InvalidArgumentException(HERE) << "Error: cannot draw cloud based on empty data.";
+  Graph graph(title, "Input marginal number", "", true, "");
+
+  /* Define cloud */
+  Sample data(values.getDimension(), 2);
+  for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
+  {
+    data(k, 0) = k + 1.0;
+    data(k, 1) = values[k];
+  }
+
+  Cloud cloud(data, "red", "circle", "");
+  graph.add(cloud);
+
+  /* Min & max indices values */
+  const Scalar minInd = data.getMin()[1];
+  const Scalar maxInd = data.getMax()[1];
+
+  /* Add text description */
+  for (UnsignedInteger k = 0; k < values.getDimension(); ++k)
+  {
+    data(k, 0) = (k + 1.0) + 0.08;
+    data(k, 1) = values[k];
+  }
+
+  Description names(values.getDimension());
+  for(UnsignedInteger i = 0; i < values.getDimension(); ++i)
+  {
+    OSS oss;
+    oss << "X" << i + 1;
+    names[i] = String(oss);
+  }
+
+  Text text(data, names, "right");
+  text.setColor("black");
+  graph.add(text);
+
+  // Set bounding box
+  Point lowerBound(2);
+  lowerBound[0] = 0.0;
+  if (minInd < 0)
+    lowerBound[1] = 1.1 * minInd;
+  else
+    lowerBound[1] = 0.9 * minInd;
+
+  Point upperBound(2);
+  upperBound[0] = values.getDimension() + 1.0 ;
+
+  if (maxInd > 0)
+    upperBound[1] = 1.1 * maxInd;
+  else
+    upperBound[1] = 0.9 * maxInd;
+
+  graph.setBoundingBox(Interval(lowerBound, upperBound));
+
+  return graph;
+}
+
+
+/** Draw the HSIC indices */
+Graph HSICEstimatorImplementation::drawHSICIndices() const
+{
+  String title = "HSIC indices";
+  Graph graph = drawValues(HSIC_XY_, title);
```

Comment:
```suggestion
  Graph graph = drawValues(getHSICIndices(), title);
```
This implements @JPelamatti's suggestion:
> Would it make sense to call the various `get` methods in the corresponding `draw` methods in order to ensure the required values have already been computed?

Reply:
Done